pkg/engine/lifecycle_test.go

```diff
@@ -5216,6 +5216,199 @@ func TestSingleResourceDefaultProviderGolangLifecycle(t *testing.T) {
+				},
+			}, nil
+		}),
+	}
+
+	newResource := func(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) error {
+		var res testResource
+		return ctx.RegisterResource("pkgA:m:typA", name, &testResourceInputs{
+			Foo: pulumi.String("bar"),
+		}, &res, opts...)
+	}
+
+	newComponent := func(ctx *pulumi.Context, name string, opts ...pulumi.ResourceOption) error {
+		var res testResource
+		err := ctx.RegisterComponentResource("pkgA:m:typA", name, &res, opts...)
+		if err != nil {
+			return err
+		}
+
+		var resChild testResource
+		return ctx.RegisterResource("pkgA:m:typA", name+"Child", &testResourceInputs{
+			Foo: pulumi.String("bar"),
+		}, &resChild, pulumi.Parent(&res))
+	}
+
+	program := deploytest.NewLanguageRuntime(func(info plugin.RunInfo, monitor *deploytest.ResourceMonitor) error {
+		ctx, err := pulumi.NewContext(context.Background(), pulumi.RunInfo{
+			Project:     info.Project,
+			Stack:       info.Stack,
+			Parallel:    info.Parallel,
+			DryRun:      info.DryRun,
+			MonitorAddr: info.MonitorAddress,
+		})
+		assert.NoError(t, err)
+
+		return pulumi.RunWithContext(ctx, func(ctx *pulumi.Context) error {
+			// Scenario #1 - apply a transformation to a CustomResource
+			res1Transformation := func(args *pulumi.ResourceTransformationArgs) *pulumi.ResourceTransformationResult {
+				// TODO[pulumi/pulumi#3846] We should use a mergeOptions-style API here.
+				return &pulumi.ResourceTransformationResult{
+					Props: args.Props,
+					Opts:  append(args.Opts, pulumi.AdditionalSecretOutputs([]string{"output"})),
+				}
+			}
+			assert.NoError(t, newResource(ctx, "res1",
+				pulumi.Transformations([]pulumi.ResourceTransformation{res1Transformation})))
+
+			// Scenario #2 - apply a transformation to a Component to transform it's children
+			res2Transformation := func(args *pulumi.ResourceTransformationArgs) *pulumi.ResourceTransformationResult {
+				if args.Name == "res2Child" {
+					// TODO[pulumi/pulumi#3846] We should use a mergeOptions-style API here.
+					return &pulumi.ResourceTransformationResult{
+						Props: args.Props,
+						Opts:  append(args.Opts, pulumi.AdditionalSecretOutputs([]string{"output", "output2"})),
+					}
+				}
+
+				return nil
+			}
+			assert.NoError(t, newComponent(ctx, "res2",
+				pulumi.Transformations([]pulumi.ResourceTransformation{res2Transformation})))
+
+			// Scenario #3 - apply a transformation to the Stack to transform all (future) resources in the stack
+			res3Transformation := func(args *pulumi.ResourceTransformationArgs) *pulumi.ResourceTransformationResult {
+				props := &testResourceInputs{
+					Foo: pulumi.String("baz"),
+				}
+
+				return &pulumi.ResourceTransformationResult{
+					Props: props,
+					Opts:  args.Opts,
+				}
+			}
+			assert.NoError(t, ctx.RegisterStackTransformation(res3Transformation))
+			assert.NoError(t, newResource(ctx, "res3"))
+
+			// Scenario #4 - transformations are applied in order of decreasing specificity
```

Comment:
This is consistent with other languages (I see that the comment was copied) but I find this surprising. The most specific transformations can get clobbered by the least specific. Any background on this @lukehoban @pgavlin?

Reply:
Yes - this feature is a "dependency injection" style feature, which is all about giving consumers the power to overwrite what a component or abstraction they are using does by default.  So for this feature in particular, we want to give ultimate authority to the outermost caller.